conanfile.py

```diff
@@ -1,112 +1,81 @@
 from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
 import os
-import shutil
-import re
+
 
 class MongoCxxConan(ConanFile):
     name = "mongo-cxx-driver"
-    version = "3.3.0"
+    version = "3.4.0"
     url = "http://github.com/bincrafters/conan-mongo-cxx-driver"
     description = "C++ Driver for MongoDB"
-    license = "https://github.com/mongodb/mongo-cxx-driver/blob/{0}/LICENSE".format(version)
-    settings =  "os", "compiler", "arch", "build_type"
-    options = {"shared": [True, False]}
-    default_options = {'shared': 'False'}
-    requires = 'mongo-c-driver/[~=1.9]@bincrafters/stable'
+    license = "Apache-2.0"
+    settings = "os", "compiler", "arch", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "polyfill": ["std", "boost", "mnmlstc", "experimental"]
+    }
+    default_options = {"shared": False, "fPIC": True, "polyfill": "boost"}
+    requires = "mongo-c-driver/1.16.1@bincrafters/stable"
     generators = "cmake"
 
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+
+    def config_options(self):
+        if self.settings.compiler == 'Visual Studio' and self.options.polyfill != "boost":
+            raise ConanInvalidConfiguration("For MSVC, best to use the boost polyfill")
+
+        if self.options.polyfill == "std":
+            tools.check_min_cppstd(self, "17")
+
+        if self.options.polyfill == "boost":
+            self.requires("boost_optional/1.69.0@bincrafters/stable")
+            self.requires("boost_smart_ptr/1.69.0@bincrafters/stable")
+
+        # Cannot model mnmlstc (not packaged, is pulled dynamically) or
+        # std::experimental (how to check availability in stdlib?) polyfill
+        # dependencies
+
+    def configure(self):
+        tools.check_min_cppstd(self, "11")
```

Comment:
```suggestion
        if self.settings.cppstd:
            tools.check_min_cppstd(self, "11")
```

Only if cppstd is configured. By default, it's empty.

Reply:
The [docs](https://docs.conan.io/en/latest/reference/tools.html#tools-check-min-cppstd) state the minimum cppstd is deduced from the compiler and compiler version if unset and its example has no such guard. Is this for when deduction fails?

Reply:
its not "deduced". There are 2 verification:
- verify self.settings.cppstd when declared (safer)
- if self.settings.cppstd is not declared compared to default cppstd used by the compiler. Both GCC 4.8 and 5 support most of c++11 features, but they don't use C++11 as default version, so it will fail for both versions.

Reply:
Okay, but by guarding it in your manner you are limiting the checks to only comparisons of `settings.compiler.cppstd`. No deduced checks will be performed with your guard, right?

The [docstring](https://github.com/conan-io/conan/blob/410be07a8666270ea79c0db181a48127251d1673/conans/client/tools/settings.py) reads:

```
        1. If settings.compiler.cppstd, the tool will use settings.compiler.cppstd to compare
        2. It not settings.compiler.cppstd, the tool will use compiler to compare (reading the
           default from cppstd_default)
        3. If not settings.compiler is present (not declared in settings) will raise because it
           cannot compare.
```

With your suggestion, we only call the `check_min_cppstd` if `self.settings.compiler.cppstd` is set, and so there are never any checks using compiler defaults.

Reply:
> With your suggestion, we only call the check_min_cppstd if self.settings.compiler.cppstd is set, and so there are never any checks using compiler defaults.

Because gcc4.9 and gcc5 supports c++11 but check_min_cppstd will fail. Thake a look in (this)[https://github.com/conan-io/conan-center-index/blob/52cb2a3d4aab8a3075d4b1b90bd353e7309cef25/recipes/taocpp-json/all/conanfile.py#L28] example and try to use the same idea. As c++11 is "old", you don't need to check a minimal compiler version, but `check_min_cppstd` should be used only when `self.settings.compiler.cppstd` is configured

Reply:
> Because gcc4.9 and gcc5 supports c++11 but check_min_cppstd will fail

This is confusing me greatly. In an unguarded call to `check_min_cppstd` an unspecified cppstd build with gcc 4.9 will raise because the compiler default cppstd is too low to build the project. If the cppstd is set in the settings of the build, the unguarded call passes and the minimum c++ standard version is set in the cmake code for the build, which can complete successfully.

If the call is guarded, we are falsely allowing unset minimum c++ standard builds with gcc 4.9 to go ahead and fail during the build instead of raising misconfiguration exceptions.

Why is it preferred not to check at all and let builds with old compilers error out instead of not even checking for c++11 because it's "old"? Isn't an unguarded fn call explicitly handling all potential failure cases?

What should I have done about [this build failure with gcc 4.9](https://travis-ci.com/bincrafters/conan-mongo-cxx-driver/jobs/290985621#L2120) instead? I'm not sure, but I'm expecting this build failure to come back with the current changeset

Shouldn't we be specifying the cppstd version in the CI build configs?

Reply:
> Shouldn't we be specifying the cppstd version in the CI build configs?

Yes, but it's not possible because it was introduced after Conan 1.0 which means it's a breaking behavior. That's why we need to check if it's configured.

Reply:
Okay, compatibility reasons I get. I added the `get_safe` version of fetching `cppstd` from your example. But an older conan will still choke on the import of the `check_min_cppstd` function anyway, no?

Reply:
Yes, but we assume that people are running the latest version ever.